### PR TITLE
Bookmarks

### PR DESCRIPTION
**Pull request by fredrikiselius**
Added:
* GUI components for bookmarks
* Drag and drop for both bookmarks and bookmark containers (ignores drop position)
* Context menus for bookmarks and containers (rename, remove)
* Clicking a bookmark will play the correct video at the correct frame
* Thumbnail and original image generation

Missing:
* Remove bookmark from the project. (Remove only removes the gui component)
* Drag and drop to the correct list row

### CODE REVIEW

File: ViAn/GUI/bookmarkcategory.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    QScrollArea* scroll_area = new QScrollArea();
+    scroll_area->setWidget(container);
+    scroll_area->setFrameStyle(0);
+    scroll_area->setWidgetResizable(true);
+    return scroll_area;
+}
+
+/**
+ * @brief BookmarkCategory::on_text_edited
+ * Updateds name with the text from the editbox
```
Comment (henpe071) [inline ViAn/GUI/bookmarkcategory.cpp]: Updates

File: ViAn/GUI/bookmarkcategory.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#include <QScrollArea>
+#include <QCommonStyle>
+#include <QMenu>
+#include "bookmarkwidget.h"
+#include "bookmarkcontainer.h"
+#include "bookmarklist.h"
+/**
+ * @brief The BookmarkCategory class
+ * This class is inteded to store two scrollable lists
+ * containing bookmars or images.
```
Comment (henpe071) [inline ViAn/GUI/bookmarkcategory.h]: bookmarks

File: ViAn/GUI/bookmarkcontainer.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+BookmarkContainer::BookmarkContainer(QWidget *parent) : QListWidget(parent) {
+    // Enable drag and drop
+    setSelectionMode(QAbstractItemView::SingleSelection);
+    setDragDropMode(QAbstractItemView::DragDrop);
+    setDragEnabled(true);
+    viewport()->setAcceptDrops(true);
+    setDropIndicatorShown(true);
+    layout = new QVBoxLayout();
+    setLayout(layout);
+    setIconSize(QSize(80,80));
```
Comment (henpe071) [inline ViAn/GUI/bookmarkcontainer.cpp]: 80 could probably be a constant

File: ViAn/GUI/bookmarklist.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+
+    if (drop_action == Qt::MoveAction) {
+        auto item = takeItem(row(clicked_item));
+        delete item;
+    }
+}
+
+void BookmarkList::dragEnterEvent(QDragEnterEvent *event) {
+    // Only accpt the event if the drop item is of the correct type
```
Comment (henpe071) [inline ViAn/GUI/bookmarklist.cpp]: Accept

File: ViAn/GUI/bookmarkwidget.cpp [hunk 5]
```diff
@@ -1,38 +1,138 @@
...
+    BookmarkCategory* f2 = new BookmarkCategory("name", bm_list, CONTAINER);
+    bm_list->addItem(f2);
+    connect(f2, SIGNAL(set_bookmark_video(VideoProject*,int)), this, SIGNAL(play_bookmark_video(VideoProject*,int)));
+
 }
 
-void BookmarkWidget::create_bookmark(const int frame_nbr, cv::Mat)
+BookmarkCategory* BookmarkWidget::add_to_container(BookmarkItem *bm_item, std::pair<int, string> *container) {
+    BookmarkCategory* bm_cat = nullptr;
+    for (int i = 0; i < bm_list->count(); i++) {
```
Comment (henpe071) [inline ViAn/GUI/bookmarkwidget.cpp:46]: Why not use for (auto a : bm_list) ?

File: ViAn/GUI/bookmarkwidget.cpp [hunk 6]
```diff
@@ -1,38 +1,138 @@
...
+            BookmarkCategory* _tmp_cat = dynamic_cast<BookmarkCategory*>(item);
+            if (_tmp_cat->get_name() == container->second) {
+                    bm_cat = _tmp_cat;
+                    break;
+                }
+            }
+        }
+
+    if (bm_cat == nullptr) {
+        // Container does not exist. Create and add it
```
Comment (henpe071) [inline ViAn/GUI/bookmarkwidget.cpp:58]: Needs a check for empty container name somewhere

File: ViAn/GUI/projectwidget.cpp [hunk 7]
```diff
@@ -165,8 +169,10 @@ void ProjectWidget::open_project() {
         clear();
         create_default_tree();
         m_proj = Project::fromFile(project_path.toStdString());
+        emit proj_path(m_proj->getDir());
```
Comment (henpe071) [inline ViAn/GUI/projectwidget.cpp:175]: getDir is wrong standard

File: ViAn/Project/bookmark.cpp [hunk 8]
```diff
@@ -56,6 +66,46 @@ void Bookmark::set_description(const std::string& text) {
...
+ * @param new_name
+ */
+void Bookmark::rename_container(std::string old_name, std::string new_name) {
+    std::for_each(m_containers.begin(), m_containers.end(),
+                  [&old_name, &new_name](std::pair<int, std::string> &c){if (c.second == old_name) c.second = new_name;});
+}
+
+/**
+ * @brief Bookmark::set_video_project
+ * Sets the video project that the bookmark belogs to
```
Comment (henpe071) [inline ViAn/Project/bookmark.cpp:93]: belongs

File: ViAn/Project/bookmark.cpp [hunk 9]
```diff
@@ -74,11 +132,22 @@ void Bookmark::read(const QJsonObject& json){
  * Writes a bookmark to a Json object, and exports the frame.
  */
 void Bookmark::write(QJsonObject& json){
+    qDebug() << "Saving";
```
Comment (henpe071) [inline ViAn/Project/bookmark.cpp]: remove

File: ViAn/Project/project.cpp [hunk 10]
```diff
@@ -213,8 +213,7 @@ std::string Project::getDir_videos() const
     return m_dir_videos;
 }
 
-std::string Project::getDir() const
-{
+std::string Project::getDir() const {
```
Comment (henpe071) [inline ViAn/Project/project.cpp:216]: fortfarande fel standard

**Review by henpe071 — COMMENTED** [review]
mostly misspelled comments, feel free to ignore

File: ViAn/GUI/bookmarkwidget.cpp (hunk 5 above)
Comment (fredrikiselius) [reply ViAn/GUI/bookmarkwidget.cpp:46]: its not supported

**Comment by henpe071** [comment]
Looks perfect

**Review by henpe071 — APPROVED** [review]
Good

**Pull request merged**